beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -4308,7 +4308,40 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
             state.latest_block_header().canonical_root()
         };
 
-        let proposer_index = state.get_beacon_proposer_index(state.slot(), &self.spec)? as u64;
+        let dependent_root = state.proposer_shuffling_decision_root(self.genesis_block_root)?;
+
+        let proposer_slot = state.slot();
+        let proposer_epoch = proposer_slot.epoch(T::EthSpec::slots_per_epoch());
+
+        let cached_proposer = self
+            .beacon_proposer_cache
+            .lock()
+            .get_slot::<T::EthSpec>(dependent_root, proposer_slot);
+
+        let proposer_index = if let Some(proposer) = cached_proposer {
+            proposer.index as u64
+        } else {
+            let (proposers, decision_root, _, fork) =
+                compute_proposer_duties_from_head(proposer_epoch, self)
+                    .map_err(BlockProductionError::BeaconChain)?;
```

Comment:
I think we could simplify this a little by calling `state.get_beacon_proposer_indices` on the state directly. It has already been advanced through to the slot of the proposal (above) so it should be able to serve the call.

The `compute_proposer_duties_from_head` basically does the same thing except it also clones the `head` again (which is slow). Using the head is also not _necessarily_ correct if we are proposing on a non-head block (sometimes in tests we propose blocks using an old parent instead of the head).

The fork can be set to `state.fork()` directly, and we also have the `dependent_root` (same as the decision root) above.

Reply:
@michaelsproul updated! But delete some proposer cache prime check in some tests to pass CI.